### ISSUE TITLE
Broadcasted convolution masks

### ISSUE THREAD

**Issue by psmaragdis** [deepmind/dm-haiku#457]
For the convolution functions, the mask argument has to match the shape of the weights over all the dimensions.  It would be very useful to instead allow broadcasting for the mask application so that one can specify a mask over all filters (or other dimensions).

As an example, suppose I wanted to add a simple tapering window.  Ideally I should do:

```
hk.Conv1D(
    output_channels=output_channels, 
    kernel_shape=kernel_size, 
    mask=jnp.hanning( kernel_size)[:,None,None]
)
```

Which unfortunately raises an exception since the weights and mask are different sizes.  In order for this to work I need to keep track of the input channels (which Haiku conveniently doesn't otherwise need), and instead use the more cumbersome:

```
hk.Conv1D( 
    output_channels=output_channels, 
    kernel_shape=kernel_size,
    mask=jnp.tile( jnp.hanning( kernel_size)[:,None,None], (1,output_channels,input_channels))
)
```

At a cursory glance, this could be simply solved by changing the mask/weights size check to make sure that the two shapes are broadcastable, and not necessarily exactly the same.
